File: typescript/src/transport.ts
/**
 * Transport layer interfaces — protocol-agnostic contracts.
 * Implementations: WsTransport, HttpTransport, etc.
 */

/** Single bidirectional message channel (one per connected peer). */
export interface ITransportConnection {
  send(raw: string): void;
  close(): void;
  readonly isOpen: boolean;
  onMessage: ((raw: string) => void) | null;
  onClose: (() => void) | null;
}

/** Client-side transport — manages outgoing connection lifecycle. */
export interface ITransportClient {
  connect(): void;
  disconnect(): void;
  send(raw: string): void;
  readonly connected: boolean;
  waitConnected(timeoutMs?: number): Promise<void>;
  onOpen: (() => void) | null;
  onClose: (() => void) | null;
  onMessage: ((raw: string) => void) | null;
  onAuthFailed: (() => void) | null;
}

/** Server-side transport — accepts incoming connections. */
export interface ITransportServer {
  start(): Promise<void>;
  stop(): Promise<void>;
  readonly address: { host: string; port: number } | null;
  onConnection:
    | ((conn: ITransportConnection, meta: Record<string, unknown>) => void)
    | null;
}

File: python/viberpc/transport.py
"""Transport layer protocols — protocol-agnostic contracts."""

from __future__ import annotations

from typing import Any, Awaitable, Callable, Protocol, runtime_checkable


@runtime_checkable
class ITransportConnection(Protocol):
    """Single bidirectional message channel (one per connected peer)."""

    async def send(self, raw: str) -> None: ...
    async def close(self) -> None: ...

    @property
    def is_open(self) -> bool: ...

    on_message: Callable[[str], Any] | None
    on_close: Callable[[], Any] | None


@runtime_checkable
class ITransportClient(Protocol):
    """Client-side transport — manages outgoing connection lifecycle."""

    async def connect(self, timeout: float = 10.0) -> None: ...
    async def disconnect(self) -> None: ...
    async def send(self, raw: str) -> None: ...

    @property
    def connected(self) -> bool: ...

    on_open: Callable[[], Any] | None
    on_close: Callable[[], Any] | None
    on_message: Callable[[str], Any] | None
    on_auth_failed: Callable[[], Any] | None


@runtime_checkable
class ITransportServer(Protocol):
    """Server-side transport — accepts incoming connections."""

    async def start(self) -> None: ...
    async def stop(self) -> None: ...

    @property
    def address(self) -> tuple[str, int] | None: ...

    on_connection: Callable[[ITransportConnection, dict[str, Any]], Any] | None
